omegaconf/listconfig.py

```diff
@@ -580,6 +581,7 @@ def _set_value_impl(
         elif vk is ValueKind.MANDATORY_MISSING:
             self.__dict__["_content"] = "???"
         elif vk == ValueKind.INTERPOLATION:
+            grammar_parser.parse(value)  # validate syntax
```

Comment:
I tried it and it had a higher performance hit (running full tests went from ~8s to ~10s), because `get_value_kind()` is also used on access, not just on assignment. I thought your idea of doing it on assignment was better :)

I was actually able to get rid of this performance hit by adding a `lru_cache` on the validation, which is also a potentially valid option (but then we need to decide the cache size -- I had 10K in mind).

Let me know what you prefer / if you have a better idea.

Reply:
Or, I can add a new `validate_interpolation_syntax` to `get_value_kind()` that would only be set to True when called from the assginment methods?

Reply:
Gotcha.
I think it makes sense to add a flag like strict_interpolation_validation that will enable complete validation, and to use it in the areas you are manually validating now.

Reply:
> from ~8s to ~10s

8s is quiet a lot. On my workstation, master tests takes about 4.5s.
tested your branch now and it takes 4.9s now, so not a major slowdown.

In any case, tests are not a good benchmark.
here is how you can use the existing benchmark to compare results:
run this on master. it will record a file. 
```
$ pytest benchmark/benchmark.py  --benchmark-autosave --benchmark-sort=name 
================================================================================================================================================================================ test session starts ================================================================================================================================================================================
platform linux -- Python 3.8.5, pytest-6.2.1, py-1.9.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /home/omry/dev/omegaconf
plugins: mock-3.5.1, lazy-fixture-0.6.3, benchmark-3.2.3
collected 11 items                                                                                                                                                                                                                                                                                                                                                                  

benchmark/benchmark.py ...........                                                                                                                                                                                                                                                                                                                                            [100%]
Saved benchmark data in: /home/omry/dev/omegaconf/.benchmarks/Linux-CPython-3.8-64bit/0007_2cf4e7c156662812bcc0a3e23a04b44927fafd43_20210211_004842.json
...
```

You can see that the filename starts with 0007, that's the run id.

In your branch, you can run the benchmarks again comparing to 0007 as a baseline. sort by name to get things grouped nicely:
```
$ pytest benchmark/benchmark.py  --benchmark-autosave --benchmark-compare=0007 --benchmark-sort=name
```

It makes it easier to compare a specific test with -k:
```
$ pytest benchmark/benchmark.py  --benchmark-autosave --benchmark-compare=0007 --benchmark-sort=name -k "test_omegaconf_create[large_dict]"
Comparing against benchmarks from: Linux-CPython-3.8-64bit/0007_2cf4e7c156662812bcc0a3e23a04b44927fafd43_20210211_004842.json
================================================================================================================================================================================ test session starts ================================================================================================================================================================================
platform linux -- Python 3.8.5, pytest-6.2.1, py-1.9.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /home/omry/dev/omegaconf, configfile: pyproject.toml
plugins: mock-3.5.1, lazy-fixture-0.6.3, benchmark-3.2.3
collected 11 items / 10 deselected / 1 selected                                                                                                                                                                                                                                                                                                                                     

benchmark/benchmark.py .                                                                                                                                                                                                                                                                                                                                                      [100%]
Saved benchmark data in: /home/omry/dev/omegaconf/.benchmarks/Linux-CPython-3.8-64bit/0009_a26a224a5be03c34cfcdf38bd67bab34b56c2f08_20210211_005244.json



--------------------------------------------------------------------------------------------------- benchmark: 2 tests --------------------------------------------------------------------------------------------------
Name (time in ms)                                         Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_omegaconf_create[large_dict] (0007_2cf4e7c)     428.0800 (1.38)     472.4565 (1.33)     451.3789 (1.36)     16.8862 (1.0)      453.2212 (1.33)     24.0847 (1.0)           2;0  2.2154 (0.74)          5           1
test_omegaconf_create[large_dict] (NOW)              310.9464 (1.0)      356.3601 (1.0)      331.8746 (1.0)      19.7616 (1.17)     339.6214 (1.0)      32.5840 (1.35)          3;0  3.0132 (1.0)           5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

The most important column is the mean.
You can see that the mean time for creating a large dict in the test went from 472ms to 356ms, which means things are actually faster now.
I think this is because we removed the regex match attempt to detect interpolations by string contains followed by a more expensive test.

This is what it looks like with pretty colors:
![image](https://user-images.githubusercontent.com/376455/107592106-f8620b80-6bc0-11eb-914e-7fec11001732.png)

Reply:
> I think it makes sense to add a flag like strict_interpolation_validation that will enable complete validation, and to use it in the areas you are manually validating now.

Done in 9fe5f53

Reply:
> here is how you can use the existing benchmark to compare results:

Thanks for the tips -- I had used a bit the benchmark before, but I didn't know about these commands.